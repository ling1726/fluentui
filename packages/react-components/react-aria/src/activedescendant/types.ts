import * as React from 'react';

export interface ActiveDescendantImperativeRef {
  find: (pred: (id: string) => boolean) => string | undefined;
  first: () => void;
  last: () => void;
  next: () => void;
  prev: () => void;
  blur: () => void;
  active: () => string | undefined;
  focus: (id: string) => void;
}

export interface ActiveDescendantOptions {
  matchOption: (el: HTMLElement) => boolean;
  imperativeRef?: React.RefObject<ActiveDescendantImperativeRef>;
}
