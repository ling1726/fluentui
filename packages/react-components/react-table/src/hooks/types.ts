import { SortDirection } from '../components/Table/Table.types';
import { TableHeaderCellProps } from '../components/TableHeaderCell/TableHeaderCell.types';

export type RowId = string | number;
export type ColumnId = string | number;
export type GetRowIdInternal<TItem> = (rowId: TItem, index: number) => RowId;

export interface ColumnDefinition<TItem> {
  columnId: ColumnId;
  compare?: (a: TItem, b: TItem) => number;
}

export interface SortStateInternal<TItem> {
  sortDirection: SortDirection;
  sortColumn: ColumnId | undefined;
  toggleSort: (columnId: ColumnId) => void;
  headerSortProps: (columnId: ColumnId) => TableHeaderCellProps;
  /**
   * Returns a sorted **shallow** copy of original items
   */
  sort: (items: TItem[]) => TItem[];
}

export interface UseTableOptions<TItem> {
  columns: ColumnDefinition<TItem>[];
  items: TItem[];
  selectionMode?: 'single' | 'multiselect';
  getRowId?: (item: TItem) => RowId;
}

export interface SelectionStateInternal {
  clearSelection: () => void;
  deSelectRow: (rowId: RowId) => void;
  selectRow: (rowId: RowId) => void;
  toggleSelectAllRows: () => void;
  toggleRowSelect: (rowId: RowId) => void;
  selectedRows: Set<RowId>;
  allRowsSelected: boolean;
  someRowsSelected: boolean;
}

export interface SortState {
  sortDirection: SortDirection;
  sortColumn: ColumnId | undefined;
  toggleSort: (columnId: ColumnId) => void;
  headerSortProps: (columnId: ColumnId) => TableHeaderCellProps;
}

export interface SelectionState {
  /**
   * Clears all selected rows
   */
  clearSelection: () => void;
  /**
   * Selects single row
   */
  selectRow: (rowId: RowId) => void;
  /**
   * De-selects single row
   */
  deSelectRow: (rowId: RowId) => void;
  /**
   * Toggle selection of all rows
   */
  toggleSelectAllRows: () => void;
  /**
   * Toggle selection of single row
   */
  toggleRowSelect: (rowId: RowId) => void;
  /**
   * Collection of row ids corresponding to selected rows
   */
  selectedRows: RowId[];
  /**
   * Whether all rows are selected
   */
  allRowsSelected: boolean;
  /**
   * Whether some rows are selected
   */
  someRowsSelected: boolean;
}

export interface RowState<TItem> {
  /**
   * User provided data
   */
  item: TItem;
  /**
   * Toggle the selection of the row
   */
  toggleSelect: () => void;
  /**
   * Selects the row
   */
  selectRow: () => void;
  /**
   * De-selects the row
   */
  deSelectRow: () => void;
  /**
   * Whether the row is selected
   */
  selected: boolean;
  /**
   * The row id, defaults to index position in the collection
   */
  rowId: RowId;
}

export interface TableState<TItem> {
  /**
   * Gets the row data for rendering
   */
  rows: () => RowState<TItem>[];
  /**
   * State and actions to manage row selection
   */
  selection: SelectionState;
  /**
   * State and actions to manage row sorting
   */
  sort: SortState;
}
