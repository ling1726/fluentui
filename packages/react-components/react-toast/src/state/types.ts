import { EVENTS } from './constants';

export type ToastId = string;

export type ToastPosition = 'top-right' | 'top-center' | 'top-left' | 'bottom-right' | 'bottom-center' | 'bottom-left';

export interface ToastOptions {
  toastId?: ToastId;
  position?: ToastPosition;
  content?: unknown;
  timeout?: number;
  pauseOnWindowBlur?: boolean;
  pauseOnHover?: boolean;
}

export interface ToasterOptions extends DefaultToastOptions {}

export interface DefaultToastOptions
  extends Pick<ToastOptions, 'position' | 'timeout' | 'pauseOnWindowBlur' | 'pauseOnHover'> {}

export interface ValidatedToastOptions extends Required<DefaultToastOptions> {}

export interface Toast extends Required<ToastOptions> {
  close: () => void;
  remove: () => void;
}

export interface DismissToastEventDetail {
  toastId: ToastId | undefined;
}

export interface ToastEventMap {
  [EVENTS.show]: CustomEvent<ToastOptions>;
  [EVENTS.dismiss]: CustomEvent<DismissToastEventDetail>;
}

export type ToastEventListenerGeneric<K extends keyof ToastEventMap> = (e: ToastEventMap[K]) => void;
export type ToastEventListener = <K extends keyof ToastEventMap>(e: ToastEventMap[K]) => void;
