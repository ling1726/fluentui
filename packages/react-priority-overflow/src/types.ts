import type { OverflowGroupState, OverflowItemEntry } from '@fluentui/priority-overflow';
import * as React from 'react';
export interface OverflowContextValue {
  itemVisibility: Record<string, boolean>;
  groupVisibility: Record<string, OverflowGroupState>;
  hasOverflow: boolean;
  registerItem: (item: OverflowItemEntry) => () => void;
  updateOverflow: (padding?: number) => void;
}

export interface UseOverflowContainerReturn<TElement extends HTMLElement> {
  /**
   * Ref to apply to the container that will overflow
   */
  containerRef: React.RefObject<TElement>;
  /**
   * Registers and overflow item
   */
  registerItem: OverflowContextValue['registerItem'];
  /**
   * Imperative function to trigger overflow update
   */
  updateOverflow: OverflowContextValue['updateOverflow'];
}
