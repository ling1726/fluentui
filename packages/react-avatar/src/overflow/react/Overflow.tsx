import * as React from 'react';
import { useMergedRefs } from '@fluentui/react-utilities';
import { OverflowGroupState, OnUpdateOverflow, OverflowEventPayload } from '../native/overflowManager';
import { OverflowContext } from './overflowContext';
import { useOverflowContainer, UseOverflowContainerOptions } from './useOverflowContainer';

export type OverflowChildProps = React.HTMLAttributes<HTMLElement> & {
  ref?: React.Ref<never>;
};

export type OverflowProps = UseOverflowContainerOptions & {
  children: React.ReactElement;
};

export const Overflow = React.forwardRef((props: OverflowProps, ref) => {
  const [hasOverflow, setHasOverflow] = React.useState(false);
  const [itemVisibility, setItemVisibility] = React.useState<Record<string, boolean>>({});
  const [groupVisibility, setGroupVisibility] = React.useState<Record<string, OverflowGroupState>>({});

  const updateItemVisibility: OnUpdateOverflow = (data: OverflowEventPayload) => {
    setHasOverflow(() => data.invisibleItems.length > 0);
    setItemVisibility(() => {
      const newState: Record<string, boolean> = {};
      data.visibleItems.forEach(x => (newState[x.id] = true));
      data.invisibleItems.forEach(x => (newState[x.id] = false));
      return newState;
    });
    setGroupVisibility(groupVisibility);
  };

  const { containerRef, registerItem, updateOverflow } = useOverflowContainer(updateItemVisibility, {
    ...props,
  });

  const mergedRef = useMergedRefs(containerRef, ref);

  const children = React.cloneElement(React.Children.only(props.children), { ref: mergedRef });

  return (
    <OverflowContext.Provider
      value={{
        itemVisibility,
        groupVisibility,
        hasOverflow,
        registerItem,
        updateOverflow,
      }}
    >
      {children}
    </OverflowContext.Provider>
  );
});
