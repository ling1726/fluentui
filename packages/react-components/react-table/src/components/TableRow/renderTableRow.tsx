import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { TableRowState, TableRowSlots } from './TableRow.types';
import { TableRowSubtleContextProvider } from '../../contexts/tableRowSubtleContext';

/**
 * Render the final JSX of TableRow
 */
export const renderTableRow_unstable = (state: TableRowState) => {
  const { slots, slotProps } = getSlots<TableRowSlots>(state);

  return (
    <TableRowSubtleContextProvider value={state.renderSubtle}>
      <slots.root {...slotProps.root} />
    </TableRowSubtleContextProvider>
  );
};
