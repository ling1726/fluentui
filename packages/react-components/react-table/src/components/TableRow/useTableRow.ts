import * as React from 'react';
import { getNativeElementProps, mergeCallbacks, useEventCallback, useMergedRefs } from '@fluentui/react-utilities';
import type { TableRowProps, TableRowState } from './TableRow.types';
import { useTableContext } from '../../contexts/tableContext';
import { useFocusVisible } from '@fluentui/react-tabster';

/**
 * Create the state required to render TableRow.
 *
 * The returned state can be modified with hooks such as useTableRowStyles_unstable,
 * before being passed to renderTableRow_unstable.
 *
 * @param props - props from this instance of TableRow
 * @param ref - reference to root HTMLElement of TableRow
 */
export const useTableRow_unstable = (props: TableRowProps, ref: React.Ref<HTMLElement>): TableRowState => {
  const { noNativeElements, size } = useTableContext();
  const rootComponent = props.as ?? noNativeElements ? 'div' : 'tr';
  const focusVisibleRef = useFocusVisible();
  const [renderSubtle, setRenderSubtle] = React.useState(false);

  const onMouseEnter = (e: React.MouseEvent<HTMLDivElement>) => {
    setRenderSubtle(true);
  };

  const onMouseLeave = (e: React.MouseEvent<HTMLDivElement>) => {
    setRenderSubtle(false);
  };

  const onFocus = (e: React.FocusEvent) => {
    if (e.currentTarget.contains(e.target)) {
      setRenderSubtle(true);
    }
  };

  const onBlur = (e: React.FocusEvent) => {
    if (!e.currentTarget.contains(e.relatedTarget)) {
      setRenderSubtle(false);
    }
  };

  return {
    components: {
      root: rootComponent,
    },
    root: getNativeElementProps(rootComponent, {
      ref: useMergedRefs(ref, focusVisibleRef),
      role: rootComponent === 'div' ? 'row' : undefined,
      onMouseEnter: useEventCallback(mergeCallbacks(onMouseEnter, props.onMouseEnter)),
      onMouseLeave: useEventCallback(mergeCallbacks(onMouseLeave, props.onMouseLeave)),
      onBlur: useEventCallback(mergeCallbacks(onBlur, props.onBlur)),
      onFocus: useEventCallback(mergeCallbacks(onFocus, props.onFocus)),
      ...props,
    }),
    size,
    noNativeElements,
    renderSubtle,
    appearance: props.appearance ?? 'none',
  };
};
