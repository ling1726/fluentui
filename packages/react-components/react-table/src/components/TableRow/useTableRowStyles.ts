import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import { tokens } from '@fluentui/react-theme';
import type { TableRowSlots, TableRowState } from './TableRow.types';
import type { SlotClassNames } from '@fluentui/react-utilities';

export const tableRowClassName = 'fui-TableRow';
export const tableRowClassNames: SlotClassNames<TableRowSlots> = {
  root: tableRowClassName,
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    display: 'flex',
    ...shorthands.padding('12px', '0px'),
    color: tokens.colorNeutralForeground1,
    ':hover': {
      backgroundColor: tokens.colorNeutralBackground1Hover,
    },
  },
});

/**
 * Apply styling to the TableRow slots based on the state
 */
export const useTableRowStyles_unstable = (state: TableRowState): TableRowState => {
  const styles = useStyles();
  state.root.className = mergeClasses(tableRowClassNames.root, styles.root, state.root.className);

  return state;
};
