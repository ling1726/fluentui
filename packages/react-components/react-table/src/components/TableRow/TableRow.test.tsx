import * as React from 'react';
import { render } from '@testing-library/react';
import { TableRow } from './TableRow';
import { isConformant } from '../../common/isConformant';
import { TableRowProps } from './TableRow.types';

const tbody = document.createElement('tbody');
describe('TableRow', () => {
  isConformant({
    Component: TableRow as React.FunctionComponent<TableRowProps>,
    renderOptions: {
      container: document.body.appendChild(tbody),
    },
    displayName: 'TableRow',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<TableRow>Default TableRow</TableRow>, { container: tbody });
    expect(result.container).toMatchSnapshot();
  });
});
