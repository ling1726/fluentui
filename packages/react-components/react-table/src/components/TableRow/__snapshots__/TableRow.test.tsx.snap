// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow renders a default state 1`] = `
<tbody>
  <tr
    class="fui-TableRow"
  >
    Default TableRow
  </tr>
</tbody>
`;
